[Runtime][IRGen] Trap C++ exceptions on *throw*, not catch.

al45tair opened the pull request:
The previous approach was effectively to catch the exception and then run a trap instruction.  That has the unfortunate feature that we end up with a crash at the catch site, not at the throw site, which leaves us with very little information about which exception was thrown or where from.

(Strictly we do have the exception pointer and could obtain exception information, but it still won't tell us what threw it.)

Instead of that, set a personality function for Swift functions that call potentially throwing code, and have that personality function trap the exception during phase 1 (i.e. *before* the original stack has been unwound).

rdar://120952971

al45tair committed:
[Runtime][IRGen] Trap C++ exceptions on *throw*, not catch.

The previous approach was effectively to catch the exception and then
run a trap instruction.  That has the unfortunate feature that we end
up with a crash at the catch site, not at the throw site, which leaves
us with very little information about which exception was thrown or
where from.

(Strictly we do have the exception pointer and could obtain exception
information, but it still won't tell us what threw it.)

Instead of that, set a personality function for Swift functions that
call potentially throwing code, and have that personality function
trap the exception during phase 1 (i.e. *before* the original stack
has been unwound).

rdar://120952971

al45tair commented:
We might need to do something to make sure that the Linux build still works for people while we're landing this — building with host tools is probably going to cause a problem because of the extra API that the compiler is using.

al45tair commented:
@swift-ci Please test

mikeash commented:
Someday, can we emit this unconditionally to prevent people trying to throw through Swift frames and catch them on the other side?

al45tair committed:
[Test] Fix tests to account for the new personality function.

The exception handling tests should check for the new personality, and
the ABI check needs to know about the new entry point in the runtime.

rdar://120952971

al45tair commented:
@swift-ci Please test

al45tair commented:
There's also an issue in that gating on availability means the test needs to pass the availability flags. I'm going to fix the availability situation on Linux and then come back to this PR.

al45tair committed:
[Runtime][Test] Add a leading underscore to swift_exceptionPersonality.

The other new runtime functions appear to have a leading underscore.
It makes sense in this case because we don't expect anything to call
this directly (other than the unwinder).

rdar://120952971

al45tair committed:
[IRGen][Test] Use the new availability support.

We need to only generate references to `_swift_exceptionPersonality`
if we're building for a new enough runtime.  The previous code was
good on Darwin, but would have resulted in build problems on Linux.

rdar://120952971

al45tair commented:
@swift-ci Please smoke test

al45tair commented:
@swift-ci Please smoke test

al45tair commented:
@swift-ci Please smoke test

al45tair commented:
Rebased.

al45tair commented:
@swift-ci Please smoke test

hyp reviewed (approved):
Thanks, that should be fine
